src/main/java/edu/cornell/kfs/fp/document/web/struts/RecurringDisbursementVoucherAction.java

```diff
@@ -0,0 +1,92 @@
+import edu.cornell.kfs.sys.CUKFSKeyConstants;
+
+public class RecurringDisbursementVoucherAction extends CuDisbursementVoucherAction implements Serializable {
+
+    private static final long serialVersionUID = -6417266354906825200L;
+    private static final org.apache.log4j.Logger LOG = org.apache.log4j.Logger.getLogger(RecurringDisbursementVoucherAction.class);
+    private static final String dot = ".";
+    
+    @Override
+    public ActionForward route(ActionMapping mapping, ActionForm form, HttpServletRequest request, HttpServletResponse response) throws Exception {
+        RecurringDisbursementVoucherForm recurringForm = (RecurringDisbursementVoucherForm) form;
+        if (getDocumentService().getByDocumentHeaderId(recurringForm.getDocument().getDocumentNumber()) == null) {
+            GlobalVariables.getMessageMap().putErrorWithoutFullErrorPath(dot + "Save", 
+                    KFSKeyConstants.ERROR_CUSTOM, "This Document needs to be saved before Submit");
+        }
+        ActionForward forward = super.route(mapping, form, request, response);
+        return forward;
+    }
+    
+    public ActionForward confirmAndCancel(ActionMapping mapping, ActionForm form, HttpServletRequest request, HttpServletResponse response) throws Exception {
+        RecurringDisbursementVoucherForm recurringForm = (RecurringDisbursementVoucherForm) form;
+        RecurringDisbursementVoucherDocument recurringDV = (RecurringDisbursementVoucherDocument) recurringForm.getDocument();
+        
+        if (StringUtils.isEmpty(recurringDV.getPaymentCancelReason())) {
+            GlobalVariables.getMessageMap().putError("document.paymentCancelReason", CUKFSKeyConstants.ERROR_RCDV_PAYMENT_CANCEL_REASON_REQUIRED);
+        } else {
+            Set<String> canceledPaymentGroups = getRecurringDisbursementVoucherDocumentService().cancelPDPPayments(recurringDV, recurringDV.getPaymentCancelReason());
+            if (canceledPaymentGroups == null || canceledPaymentGroups.isEmpty()) {
+                KNSGlobalVariables.getMessageList().add(CUKFSKeyConstants.INFO_RCDV_NO_PAYMENTS_CANCELED);
+            } else {
+                KNSGlobalVariables.getMessageList().add(CUKFSKeyConstants.INFO_RCDV_LIST_OF_PAYMENT_GROUPS_CANCELED, 
+                        buildCommaSpacedList(canceledPaymentGroups));
+            }
+        }
+        return mapping.findForward(RiceConstants.MAPPING_BASIC);
+    }
+    
+    private String buildCommaSpacedList(Set<String> canceledPaymentGroups) {
+        StringBuffer message = new StringBuffer();
+        int counter = 0;
+        for (String group : canceledPaymentGroups) {
+            if (counter > 0) {
+                message.append(", ");
+            }
+            message.append(group);
+            counter++;
+        }
+        return message.toString();
+    }
+    
+    public RecurringDisbursementVoucherDocumentService getRecurringDisbursementVoucherDocumentService() {
```

Comment:
Would it be worthwhile to add corresponding setters here and convert the getters to the if (service == null) {service =  SpringContext.getBean()} return service ideom?